activerecord/lib/active_record/attribute_methods/primary_key.rb

```diff
@@ -31,6 +31,7 @@ def primary_key_values_present? # :nodoc:
       # Sets the primary key column's value.
       def id=(value)
         if self.class.composite_primary_key?
+          raise TypeError, "Expected value matching #{self.class.primary_key.inspect}, got #{value.inspect}." unless value.is_a?(Enumerable)
```

Comment:
Wonder if we should be even more restrictive and check for the enumerable's `size` so we also raise if the value is smaller or larger than the primary key.

Reply:
I considered this as well, but decided not to because I'm not sure of the performance implications. I'll take a look at benchmarking it to see if checking the length of the PK impacts much.